src/components/Table/Table.tsx

```diff
@@ -1,27 +1,94 @@
-import { Table as MantineTable, TextInput } from "@mantine/core";
-import { type ChangeEvent, type ReactNode, useMemo, useState } from "react";
-import { IconSortAscending, IconSortDescending } from "@tabler/icons-react";
+import {
+  Table as MantineTable,
+  TextInput,
+  Pagination as MantinePagination,
+  createStyles,
+  Box,
+} from "@mantine/core";
+
+import {
+  type ChangeEvent,
+  type ReactNode,
+  useEffect,
+  useMemo,
+  useState,
+} from "react";
+
+import {
+  IconArrowBarToLeft,
+  IconArrowBarToRight,
+  IconArrowLeft,
+  IconArrowRight,
+  IconGripHorizontal,
+  IconSortAscending,
+  IconSortDescending,
+} from "@tabler/icons-react";
+
 import { omit } from "lodash/fp";
-import { filterData, getNextSortValue, sortData } from "../../utils/utils.ts";
+
+import {
+  filterData,
+  getMaximumPages,
+  getNextSortValue,
+  sortData,
+} from "../../utils/utils.ts";
+
+import { useElementSize } from "@mantine/hooks";
 
 export interface Column<TData> {
   columnName: string;
   exactMatch: boolean;
   header: string;
   cellRenderer?: (row: TData) => ReactNode;
+  isFilterable: boolean;
 }
 
 interface TableProps<TData> {
   data: TData[];
   columns: Array<Column<TData>>;
+  rowHeight: number;
 }
 
 export type Sort = Record<string, "asc" | "desc" | null>;
 export type Filter = Record<string, string>;
 
+const DEBOUNCE_DELAY = 20;
+
+const TABLE_PADDING = 50;
 export function Table<T extends { id: number }>(props: TableProps<T>) {
+  const { classes } = useStyles();
+
+  const { ref: tableRef, height: tableHeight } = useElementSize();
+  const { ref: paginationRef, height: paginationHeight } = useElementSize();
+
+  const [adjustedTableHeight, setAdjustedTableHeight] = useState(tableHeight);
```

Comment:
Why do we need to store in state the adjustedHeight? Same waterfall applies here.

Think about it this way: you're adding an unnecessary re-render, just to store the value. Sometimes it means the component will jump after first re-render.

`adjustedHeight` is a derivative of various calculations. If you've calculated them, then simply keep using it for further stuff, instead of setting state with the new value, and then re-calculating stuff requiring it.